feat(contianers): aggregated deployment readines

lusergit opened the pull request:
`ready?` calculation on deployment resources is more descriptful, reporting, when not ready, the non-ready resources.

deployment campaigns when relying on the backend to compute the readyness of a deployment (and mark the target as successful/failed/etc.) can have a more descriptive view of what's happening, with missing resources to handle.

<!--

**Please, carefully describe what the PR does and why you are opening it.**

Short check list:

* [ ] Please, make sure to read CONTRIBUTING.md and CODE_OF_CONDUCT.md
* [ ] Make sure to open your PR against the right branch: master / release-VERSION
* [ ] Make sure to sign-off all your commits
* [ ] GPG signing is appreciated
* [ ] Make sure the code follows coding style (use automated formatting, such as `mix format`)

-->

davidebriani committed:
feat: server side pagination (#888)

Merges `feature/server-side-pagination` branch into main.

All queries are now paginated with relay to allow the dynamic retrieval of information form the backend.

Signed-off-by: Davide Briani <davide.briani@secomind.com>

lusergit committed:
feat(contianers): aggregated deployment readines

`ready?` calculation on deployment resources is more descriptful, reporting,
when not ready, the non-ready resources.

deployment campaigns when relying on the backend to compute the readyness of a
deployment (and mark the target as successful/failed/etc.) can have a more
descriptive view of what's happening, with missing resources to handle.

Signed-off-by: Luca Zaninotto <luca.zaninotto@secomind.com>

davidebriani commented on backend/lib/edgehog/deployment_campaigns/deployment_mechanism/lazy/executor.ex:
```diff
@@ -420,19 +420,21 @@ defmodule Edgehog.DeploymentCampaigns.DeploymentMechanism.Lazy.Executor do
-        {:ready, _} -> [internal_event({:deployment_success, deployment})]
-        {_, :error} -> [internal_event({:deployment_failure, deployment})]
-        {_, _} -> []
+      case ready? do
+        :ready ->
+          [internal_event({:deployment_success, deployment})]
+
+        {:not_ready, missing} ->
+          if errors?(missing),
+            do: [internal_event({:deployment_failure, deployment})],
+            else: []
```
This block of code made me think.

Maybe we could find a better way to expose the status of the whole deployment from the Deployment resource.

It also seems a bit convoluted to build a single list of different resources, where some elements could be a `{_deployment, :error}` tuple.

Let's start from scratch and let's look at what we really need to know here in the executor:
1. if a deployment has stopped progressing because it succeeded
2. if a deployment has stopped progressing because it encountered an error
3. if a deployment is still progressing

Note that we don't need to know more than that here, like the exact resources that are still missing device acknowledgement.

Besides, my suspicion is that when rolling out a deployment we will not receive any error by the device regarding specific resources, since the only Astarte interface where the device can report errors is the `io.edgehog.devicemanager.apps.DeploymentEvent` interface.
During rollout, the device just acknowledges the received messages, so it's likely we only get errors when:
- The device signals an error about the deployment. It could be due to an error on a specific container or network, but it is advertised as a deployment error.
- The device does not reply, and we make the rollout operation time out.

On one hand we know when individual resources (including Deployment) are successfully rolled out, on the other hand we receive a single error to know when something breaks. Incidentally, this error is saved on the Deployment resource by setting its state to `:error` (+ a `:last_error_message`)

So it seems to me that unfortunately we are mixing things up a bit, since the Deployment resource could be successfully rolled out and yet we put it in an error state because another resource (e.g. Container) encountered an error during rollout. In this sense, we are using the Deployment `state` as both the state of the single Deployment resource and to report a general error state for all of its resources.

Additionally, it would be nice to start distinguishing between:
- the fact that the device acknowledges a resource definition, an operation that typically reaches a terminal state (e.g. `:pending` -> `:sent` -> `:received`)
- the fact that the resource is in a certain state on the device: e.g. `:stopped` -> `:starting` -> `:started` -> `:stopping` -> `:stopped`.
 
Given these considerations, let's go back to our three-parts goal in the Campaign executor.
We could compute that with a `rollout_state` calculation on the Deployment resource, which returns:
- `:rolled_out`, if all the resources (including Deployment) are successfully rolled out.
- `:progressing` if not all resources are rolled out, and the deployment is not reporting an error.
- `:error` if not all resources are rolled out, and the deployment is reporting an error (currently, `:state = :error`).

We could even type this well with a `RolloutState` enum struct.

To keep it simple, we probably don't need to compute list of missing resources, or have complex `:ready?` manual calculations. It probably suffices to have `:rolled_out?` calculations on resources derived from their `:state`. Ideally, as pointed out, in the future on each resource we should really have a `:rollout_state` that ultimately transitions to the terminal rollout state `:rolled_out`. And have a `:rollout_error` field on each resource instead of conflating different kind of errors in the same Deployment field `:state = :error`. In this case, the calculation that we need to use for the executor would just need to gather the RolloutState of all resources and combine them to get a summarized RolloutState.

davidebriani committed:
feature: application management (#934)

This commit introduces the groundwork for comprehensive application management system to Edgehog, enabling end-to-end container application lifecycle management including deployment campaigns.
This change is the first step for transforming Edgehog from a pure device management platform into a complete IoT application deployment and management solution.

## Major Features Added

### Core Application Management
- **Applications & Releases**: Complete CRUD operations for Applications, i.e. container sets, and their versioning with Releases
- **Container Management**: Full container lifecycle with support for images, networks, volumes, and device mappings
- **Resource Management**: Dedicated management for container resources (images, networks, volumes, image credentials)
- **Deployment System**: Comprehensive deployment orchestration with status tracking and lifecycle management

### Deployment Campaigns
- **Campaign Management**: Create and manage deployment campaigns for rolling out releases to device groups
- **Deployment Mechanism**: Intelligent deployment executor with retry logic and failure handling
- **Target Management**: Fine-grained control over deployment targets with status tracking
- **Channel System**: Generalized channel system (evolved from Update Channels) to support both updates and deployments

### Enhanced Device Integration
- **Application Status Tracking**: Real-time visibility into application states on devices
- **Container Runtime Integration**: Support for container operations, status monitoring, and resource management
- **Device Mapping Support**: Ability to map device files and resources into containers
- **Astarte Interface Integration**: New interfaces for container management, deployment commands, and status updates

## Scale of Changes

- **Backend**: 200+ new files, 15,000+ lines of code added
- **Frontend**: 100+ new components and pages, complete UI overhaul for application management
- **Database**: 45+ new migrations with comprehensive schema for application lifecycle
- **APIs**: 50+ new GraphQL queries and mutations
- **Interfaces**: 15+ new Astarte interfaces for device communication

## Architecture Highlights

### Backend Domains
- **`Edgehog.Containers`**: Core domain for application, release, container, and deployment management
- **`Edgehog.DeploymentCampaigns`**: Campaign orchestration with deployment mechanism
- **`Edgehog.Campaigns`**: Generalized campaign infrastructure supporting both updates and deployments

### Frontend Structure
- **Applications Section**: Complete application management UI with navigation between apps, releases, and containers
- **Deployment Management**: Campaign creation, monitoring, and target management interfaces
- **Resource Management**: Dedicated pages for networks, volumes, images, and credentials
- **Enhanced Device Views**: Application tabs showing deployed applications and their status

### Device Communication
- **Bidirectional Communication**: Request/response patterns for deployment operations
- **Status Monitoring**: Real-time updates on deployment progress and application health  
- **Resource Provisioning**: Automated creation and management of container resources on devices

## Technical Implementation

### Database Schema
- Multi-tenant architecture with proper isolation
- Cascading relationships ensuring data consistency
- Resource caching for improved performance

### GraphQL API
- Relay-style pagination for all list operations
- Strongly typed schema with comprehensive validation
- Optimized queries with proper relationship loading

### Frontend Architecture
- Relay for GraphQL data management
- Infinite scrolling tables with server-side pagination
- Comprehensive internationalization support
- Reusable component library for consistent UX

## Breaking Changes

- **Pagination**: GraphQL API now uses Relay/keyset pagination for unbounded lists
- **Channel System**: Update Channels have been generalized to support deployment campaigns
- **UI Navigation**: Significant changes to navigation structure to accommodate new features

## Migration Path

- Existing update campaigns continue to work with the generalized channel system
- UI maintains backward compatibility for existing workflows
- Astarte interfaces are additive and don't break existing functionality
